src/sdk/namespace_fs.js

```diff
@@ -1964,69 +1974,81 @@ class NamespaceFS {
                 throw new S3Error(S3Error.InvalidObjectStorageClass);
             }
 
-            // Total 8 states
+            const now = new Date();
+
+            // Total 6 states (restore_request - 2, restore_expiry - 3)
             const restore_request = stat.xattr[GlacierBackend.XATTR_RESTORE_REQUEST] ?
                 parseInt(stat.xattr[GlacierBackend.XATTR_RESTORE_REQUEST], 10) : null;
-            const restore_ongoing = stat.xattr[GlacierBackend.XATTR_RESTORE_ONGOING] === 'true';
             const restore_expiry = stat.xattr[GlacierBackend.XATTR_RESTORE_EXPIRY] ?
                 new Date(stat.xattr[GlacierBackend.XATTR_RESTORE_EXPIRY]) : null;
 
-            // 5 Valid States
-            if (restore_request) {
-                // It is possible that the restore request was made but we haven't yet
-                // started the restore process. Batching is an implementation detail
-                // and need not be leaked to the client.
-                if (!restore_ongoing && !restore_expiry) {
-                    dbg.log0('namespace_fs.restore_object: state - (restore_request, !restore_ongoing, !restore_expiry)');
-                    throw new S3Error(S3Error.RestoreAlreadyInProgress);
-                }
-            } else if (restore_ongoing) {
-                if (!restore_expiry) {
-                    dbg.log0('namespace_fs.restore_object: state - (!restore_request, restore_ongoing, !restore_expiry)');
-                    throw new S3Error(S3Error.RestoreAlreadyInProgress);
-                }
-            } else {
-                const now = new Date();
+            // 2 states
+            if (
+                (restore_request && !restore_expiry) ||
+                (restore_request && restore_expiry <= now)
```

Comment:
Do we need to differentiate between undefined and 0 for restore_request? Right now 0 behaves just like undefined, and if we need to differentiate we can check restore_request > 0 vs. >=0.